### PR TITLE
Retry process routes when name is "unknown"

### PR DESCRIPTION
**Pull request by jander-msft**
I created #523 to so that we can better mitigate this type of problem. We can never guarantee that we get all of the process information on the first call, but once we getting it, we should be able to cache it.

### CODE REVIEW

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs [hunk 1]
```diff
@@ -245,7 +245,15 @@ private void VerifyProcessInfoEquality(List<ProcessInfo> processInfos)
...
-            Assert.Single(processInfoNames.Distinct(StringComparer.Ordinal));
+            // In .NET 5+, the process name comes from the command line from the ProcessInfo command, which can fail
+            // or be abandoned if it takes too long to response. In .NET Core 3.1, the process name comes from the
+            // command line from issuing a very brief event source trace, which can also fail or be abandoned if it
+            // takes too long to finish. Additionally, for 'connect' mode, these values are recomputed for every http
+            // invocation, so a prior invocation could succeed whereas a subsequent one could fail. Much of this could
+            // be mitigated if process information could be cached between calls. In any of these scenarios, if the
+            // process name is failed to be determined, the http api will return "unknown". For testing purposes,
+            // filter out the "unknown" value before comparing the real process names. 
+            Assert.Single(processInfoNames.Except(new[] { "unknown" }, StringComparer.Ordinal).Distinct(StringComparer.Ordinal));
```
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs:256]: I actually pulled and played around with the tests for a little while because they've been improved significantly compared to what I initially wrote, but I think I found an edge case where this _might_ be an issue. Basically, if every member of processInfoNames in VerifyProcessInfoEquality happens to be unknown due to this failure to respond quickly, the Assert.Single will fail since it'll be empty (all the unknown names would have been removed). Since I can't reliably recreate this situation and force processes to have the name unknown, I just switched "unknown" to "dotnet" and the test failed. I realize this is probably a pretty limited edge case, but I figured I should mention it.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs [hunk 2]
```diff
@@ -245,7 +245,15 @@ private void VerifyProcessInfoEquality(List<ProcessInfo> processInfos)
 
             Assert.Single(processInfoPIDs.Distinct());
             Assert.Single(processInfoUIDs.Distinct());
-            Assert.Single(processInfoNames.Distinct(StringComparer.Ordinal));
+            // In .NET 5+, the process name comes from the command line from the ProcessInfo command, which can fail
+            // or be abandoned if it takes too long to response. In .NET Core 3.1, the process name comes from the
```
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs:249]: Just a typo: response -> respond

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs (hunk 1 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs:256]: Absolutely correct. I thought about this case too, but I figure its okay to ignore for now. It would require retrying until we got non-"unknown" values for each of the calls. However, it's unlikely that all three will have "unknown" for the name. When we fix #523, then we only technically have to loop for the first call and then all subsequent calls should have the same values because they'll all be cached.

Because of how infrequent this should be an issue in the test with this workaround and we should fix the non-caching issue, I figure it's okay to ignore for now. We'll reassess if the test fails frequently. I've only been making these targeted workarounds because it's had a high frequency of failure.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs (hunk 2 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs:249]: Thanks! I've been doing too much HTTP and IPC work recently (lots of "response" everywhere).

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs (hunk 1 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs:258]: I added more to this comment explaining the condition under which it can fail still but that it is unlikely to fail.

Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/ProcessTests.cs]: This ended up "kicking the can down the road" such that when we go to do the `/process` invocation with the name parameter set to "unknown" it would basically fail each time. I ended up adding try logic to all of the positive lookups in order to get process information that have non-`unknown` process names.

**Review by kkeirstead — APPROVED** [review]

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClientExtensions.cs [hunk 3]
```diff
@@ -31,6 +34,45 @@ public static async Task<IEnumerable<ProcessIdentifier>> GetProcessesAsync(this
...
+            int attempt = 0;
+            while (attempt < maxAttempts)
+            {
+                attempt++;
+
+                outputHelper.WriteLine($"Attempt #{attempt} of {maxAttempts}: GET /processes");
+
+                identifiers = (await client.GetProcessesAsync()).ToList();
+
+                if (null != filter)
```
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClientExtensions.cs:53]: Don't you want to put the filtering after the retries, since filtering by process name will be broken? (Although I didn't see any clients that do this)

**Review by wiktork — APPROVED** [review]

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClientExtensions.cs (hunk 3 above)
Comment (jander-msft) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClientExtensions.cs:53]: I wanted to filter to processes that we cared about rather than verifying that all of the processes we found have valid names. I can change the filter to just be on pid (or provide a list of pods instead of a filter) to force the intention rather that leaving it up to the tests.
